lib/Sema/NameBinding.cpp

```diff
@@ -902,55 +917,71 @@ void NameBinder::crossImport(ModuleDecl *M, UnboundImport &I) {
     // FIXME: Should we warn if M doesn't reexport underlyingModule?
     SF.addSeparatelyImportedOverlay(M, I.getUnderlyingModule().get());
 
-  // FIXME: Most of the comparisons we do here are probably unnecessary. We
-  // only need to findCrossImports() on pairs where at least one of the two
-  // modules declares cross-imports, and most modules don't. This is low-hanging
-  // performance fruit.
-
   auto newImports = crossImportableModules.getArrayRef()
                         .slice(nextModuleToCrossImport);
+
+  if (newImports.empty())
+    // Nothing to do except crash when we read past the end of
+    // crossImportableModules in that assert at the bottom.
+    return;
+
   for (auto &newImport : newImports) {
     if (!canCrossImport(newImport))
       continue;
 
-    // Search imports up to, but not including or after, `newImport`.
+    // First we check if any of the imports of modules that have declared
+    // cross-imports have declared one with this module.
+    findCrossImportsInLists(I, crossImportDeclaringModules, {newImport},
+                            /*shouldDiagnoseRedundantCrossImports=*/false);
+
+    // If this module doesn't declare any cross-imports, we're done with this
+    // import.
+    if (newImport.module.second->doesNotDeclareAnyCrossImportOverlays())
+      continue;
+
+    // Fine, we need to do the slow-but-rare thing: check if this import
+    // declares a cross-import with any previous one.
     auto oldImports =
-        make_range(crossImportableModules.getArrayRef().data(), &newImport);
-    for (auto &oldImport : oldImports) {
-      if (!canCrossImport(oldImport))
-        continue;
+        llvm::makeArrayRef(crossImportableModules.getArrayRef().data(),
+                           &newImport);
```

Comment:
We should probably annotate the arguments to indicate which overload is being picked. Replacing the `data()` call with `begin()` would help, as otherwise, one might think one is trying to pick the ptr+size overload.

Reply:
I didn't think the comments would really help here, but I added a comment explaining what we're creating.